pkg/controllers/resourcesyncer/resourcesyncer.go

```diff
@@ -0,0 +1,171 @@
+package resourcesyncer
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/utils/strings/slices"
+
+	rkev1 "github.com/rancher/rancher/pkg/apis/rke.cattle.io/v1"
+	rkecontrollers "github.com/rancher/rancher/pkg/generated/controllers/rke.cattle.io/v1"
+	corecontrollers "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	"github.com/rancher/rancher/pkg/types/config"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// resourceSyncer synchronizes secrets between upstream and downstream clusters.
+type resourceSyncer struct {
+	upstreamSecretClient   corecontrollers.SecretInterface
+	downstreamSecretClient corecontrollers.SecretInterface
+
+	rkeControlPlaneCache  rkecontrollers.RKEControlPlaneCache
+	rkeControlPlaneClient rkecontrollers.RKEControlPlaneClient
+
+	mgmtClusterName string
+	rkeControlPlane *types.NamespacedName
+}
+
+const (
+	SyncToCluster   = "provisioning.cattle.io/sync-to-cluster"
+	SyncToNamespace = "provisioning.cattle.io/sync-to-namespace"
+	SyncedTimestamp = "provisioning.cattle.io/last-synced-at"
+)
+
+// list of namespaces that are allowed to have resources synchronized between
+// the upstream and downstream cluster
+var allowedNamespaces = []string{"kube-system"}
+
+func Register(ctx context.Context, cluster *config.UserContext, mgmt *config.ScaledContext) {
+	// we only want to run this controller for non-local clusters
+	if cluster.ClusterName == "local" {
+		return
+	}
+
+	r := resourceSyncer{
+		upstreamSecretClient:   mgmt.Core.Secrets(""),
+		downstreamSecretClient: cluster.Core.Secrets(""),
+		rkeControlPlaneClient:  mgmt.Wrangler.RKE.RKEControlPlane(),
+		rkeControlPlaneCache:   mgmt.Wrangler.RKE.RKEControlPlane().Cache(),
+		mgmtClusterName:        cluster.ClusterName,
+	}
+
+	// populate the rkeControlPlane field so that it's available in the handler
+	_, err := r.fetchRkeControlPlane()
```

Comment:
I meant to leave this comment on my last review, but we could probably just unroll this function here since we only use it once, that way we don't need to store the `RKEControlPlane` client/cache.